### DISCUSSION TITLE
[question] force cache

### DISCUSSION THREAD

**Discussion by sayem314** [sindresorhus/got#1701] [Q&A] [answered]
#### What would you like to discuss?

How do I force cache requests with hooks? I have tried this and it doesn't work.

```ts
const instance = got.extend({
  cache: new Map(),
  hooks: {
    afterResponse: [
      (response) => {
        response.headers['cache-control'] = 'max-age=600';
        return response;
      },
    ],
  },
});
```

#### Checklist

- [x] I have read the documentation.

**Comment by sindresorhus** [comment, marked as answer]
The issue is that modifying the `cache-control` header in `afterResponse` doesn't affect caching because Got has already decided wether to cache the response by that point.

To force caching, you can use the `cacheOptions` to customize the cache behavior:

```javascript
import got from 'got';
import Keyv from 'keyv';

const cache = new Keyv();

const instance = got.extend({
	cache,
	cacheOptions: {
		shared: false,
		cacheHeuristic: 1, // Force heuristic caching
		immutableMinTimeToLive: 600000, // 10 minutes
		ignoreCargoCult: false
	}
});
```

Alternatively, if you want to cache everything regardless of headers, you can override the cache key logic:

```javascript
const instance = got.extend({
	cache: new Map(),
	hooks: {
		beforeRequest: [
			(options) => {
				// Force cache-friendly headers on the request
				options.headers['cache-control'] = 'public, max-age=600';
			}
		]
	}
});
```

This way Got will treat all responses as cacheable for 10 minutes.

**Reply by devadathanmb** [reply]
Is it possible to cache responses to `POST` requests this way? Or achieve similar behavior in some way?

I understand that `POST` is considered unsafe in the HTTP semantics sense (i.e., it can cause side effects and isn’t normally cacheable), and also saw some related discussion [here](https://github.com/sindresorhus/got/issues/873#issuecomment-533782782)

Basically, I’m dealing with an HTTP endpoint that uses broken semantics — it uses `POST` requests for what are essentially just read-only operations, so I want to cache those responses if possible.
